pkg/controller/direct/alloydb/instance_controller.go

```diff
@@ -0,0 +1,391 @@
+		if desired.NetworkConfig.EnablePublicIp != nil && !reflect.DeepEqual(actual.NetworkConfig.EnablePublicIp, desired.NetworkConfig.EnablePublicIp) {
+			log.V(2).Info("'spec.networkConfig.enablePublicIp' field is updated (-old +new)", cmp.Diff(actual.NetworkConfig.EnablePublicIp, desired.NetworkConfig.EnablePublicIp))
+			updatePaths = append(updatePaths, "network_config.enable_public_ip")
+		}
+		if desired.NetworkConfig.EnableOutboundPublicIp != nil && !reflect.DeepEqual(actual.NetworkConfig.EnableOutboundPublicIp, desired.NetworkConfig.EnableOutboundPublicIp) {
+			log.V(2).Info("'spec.networkConfig.enableOutboundPublicIp' field is updated (-old +new)", cmp.Diff(actual.NetworkConfig.EnableOutboundPublicIp, desired.NetworkConfig.EnableOutboundPublicIp))
+			updatePaths = append(updatePaths, "network_config.enable_outbound_public_ip")
+		}
+		if desired.NetworkConfig.AuthorizedExternalNetworks != nil && !reflect.DeepEqual(actual.NetworkConfig.AuthorizedExternalNetworks, desired.NetworkConfig.AuthorizedExternalNetworks) {
+			log.V(2).Info("'spec.networkConfig.authorizedExternalNetworks' field is updated (-old +new)", cmp.Diff(actual.NetworkConfig.AuthorizedExternalNetworks, desired.NetworkConfig.AuthorizedExternalNetworks))
+			updatePaths = append(updatePaths, "network_config.authorized_external_networks")
+		}
+	}
+	if desired.ReadPoolConfig != nil {
+		if desired.ReadPoolConfig.NodeCount != nil && !reflect.DeepEqual(actual.ReadPoolConfig.NodeCount, desired.ReadPoolConfig.NodeCount) {
+			log.V(2).Info("'spec.readPoolConfig.nodeCount' field is updated (-old +new)", cmp.Diff(actual.ReadPoolConfig.NodeCount, desired.ReadPoolConfig.NodeCount))
+			updatePaths = append(updatePaths, "read_pool_config.node_count")
+		}
+	}
+	return updatePaths, nil
+}
+
+// Export maps the GCP object to a Config Connector resource `spec`.
+func (a *instanceAdapter) Export(ctx context.Context) (*unstructured.Unstructured, error) {
+	if a.actual == nil {
+		return nil, fmt.Errorf("Find() not called")
+	}
+	u := &unstructured.Unstructured{}
+
+	obj := &krm.AlloyDBInstance{}
+	mapCtx := &direct.MapContext{}
+	obj.Spec = direct.ValueOf(AlloyDBInstanceSpec_FromProto(mapCtx, a.actual))
+	if mapCtx.Err() != nil {
+		return nil, mapCtx.Err()
+	}
+	uObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	// Split name into tokens and use ID.
+	u.SetName(a.actual.Name)
+	u.SetGroupVersionKind(krm.AlloyDBInstanceGVK)
+
+	u.Object = uObj
+	return u, nil
+}
+
+// Delete the resource from GCP service when the corresponding Config Connector resource is deleted.
+func (a *instanceAdapter) Delete(ctx context.Context, deleteOp *directbase.DeleteOperation) (bool, error) {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("deleting instance", "name", a.id)
+
+	// Returning true directly if it is to delete a secondary instance.
+	// Technically the secondary instance is only abandoned but not deleted.
```

Comment:
Yes, users can still set `cnrm.cloud.google.com/deletion-policy: "abandon"` to abandon the secondary AlloyDBInstance.

The API behavior the comment here describes basically means even if users tries to use KCC to delete a secondary AlloyDBInstance, the underlying secondary AlloyDB instance (the GCP resource) will not be deleted. This is because the instance by itself is not deletable. The deletion of a secondary AlloyDB instance is a side effect of force deletion of a secondary AlloyDB cluster, which can be done via the deletion of a secondary AlloyDBCluster with `spec.deletionPolicy` set to `FORCE` in KCC.

Reply:
Are you talking about KCC abandon but not GCP having an abandon concept?  
What does the ` a.gcpClient.DeleteInstance(ctx, req)` returns if the type is `SECONDARY`?

Reply:
Yes, I'm talking about KCC abandon. I think this GCP resource doesn't have an abandon concept.

DeleteInstance() returns the following error if it is a secondary instance:

```
googleapi: Error 400: Invalid resource state for "projects/[projectNumber]/locations/[location]/clusters/alloydbcluster-2-20241210/instances/alloydbinstance-2-20241210": Cannot delete SECONDARY instance without deleting the SECONDARY cluster. Delete the SECONDARY cluster in force mode to delete the entire cluster and the instances
```